### ISSUE TITLE
Crowdin translation environment

### ISSUE THREAD

**Issue by reindex-ot** [FreeDownloadManagerTeam/FDM6-localization#568]
Editing with PO Edit is not efficient.
I think we should use Crowdin to make progress and updates faster.
[https://crowdin.com/](https://crowdin.com/)

**Comment by comradekingu** [comment]
Consider using Weblate to avoid egregious terms and closed-source SAAS.
https://hosted.weblate.org/hosting/

Pontoon is the nearest alternative, but as a clone of Crowdin, it has the same flawed functionality.

**Comment by olavinto** [comment]
It is interesting how many find Crowdin's functionality "flawed". Personally, after actively using various platforms (Crowdin, Weblate, Transifex, OneSky...) for dozens of projects for years, I still find Crowdin's interface to be the much faster and fluid to use than anything else. It also offers various views/layouts for editing which is more than can be said of many other platforms (many users probably aren't even aware of this because they are kind of hidden behind the editor's settings button, like most other handy Crowdin features are). It is also hands down the best I've used with mobile browsers as the UI scales very well.

Weblate is fine but its interface has a lot of wasted space and could really use a lot of compaction, and it is also kind of bugged and/or lacking in some areas: the view jumps stupidly around when you click on fields (depending on your scroll position), notifications take like half the screen for some reason and its glossary is nowhere near Crowdin. Weblate is rather young in comparison so these problems are to be expected. It has pretty powerfull features that many projects could benefit from but most devs don't bother utilizing them.

Transifex is also fine but it is bloated and shoves almost everything in your face all the time. Feature wise it's mostly on par with Crowdin. But it is one of the most confusing of all I've used - not because it is difficult but because there is so much rarely used things on screen all the time. Browsing your profile's projects is also frustrating ass hell in comparison.

OneSky looks like it was made before 2010. It works but it really doesn't offer any extra features and even the basics are lacking.

Pontoon is also fine and pretty similar to Crowdin, as was said. There are also few others that are not as widely used or known but work ok.

So, unless it wasn't obvious, I also lean heavily towards Crowdin over any others. Sure Weblate is open source and may offer more from a developers standpoint and I'm also fine with it, even if I like using Crowdin a lot more.

**Comment by comradekingu** [comment]
Crowdins interface is neither quick, nor expedient.
The pagination means having to load pages and reload the view every so often.
Doing that across several thousand strings to arrive at some sort of consistency is extremely tedious,
which is why nobody does it and quality always suffers.
I could go on at length and depth, but that is the prefatory showstopper.

There are many views in Weblate. The regular one, the Zen one, the Matrix one etc.
The Zen one is hands down the quickest out of any approach.

My favourite, and where it came from is Pootle.
Transifex is also old, but its UI/UX is even more moronic than that of Crowdin.

@olavinto
Reading from your description of Crowdin is "free" in that forumpost is inaccurate.
It is only devoid of up-front monetary cost, and doesn't qualify for gratis at all, nor is it free as in freedom.
I have spent a lot of time reading terms and conditions, and people seem to forget
the only as a service model doesn't exactly bode well for privacy, rights, or anything up for grabs.

Their own language is allowing it only for non-commercial open source, which given open sourceTM doesn't allow that clause means it strictly means projects without any income.
You quickly run into Transifex style negotiations for enterprise plans once they have you over the barrel.
Very few libre software projects actually use atm. with more moving away from it, but given their practices are as shady as it comes with many documented examples, I wouldn't be surprised.
This is the level of known and possible nonsense SaaS-only grants.

If you want Crowdin functionality, use Pontoon, which unfortunately was modeled on it.

Tolgee unfortunately has unsound terms of services. Unfortunate and unenforcable.

**Comment by olavinto** [comment]
Well, personally, I have never had any problems with how Crowdin splits the strings and instead I actually like the fact that it kind of gives me a place to cut of (every 50 strings). For me the pages also tend to load faster than Weblate's Zen mode does (not rare that I have to wait for the next batch of strings to load). Zen is also exactly the one that wants to jump around because for some reason it wants to force the interface to scroll way before you get to the bottom which I find super irritating. Its column width also changes the further you get when using the side by side mode which I find it much more practical as it gives me more vertical space. For me it is much easier to read multiple shorter lines of text than fewer long (or super long depending on your screen) lines and this is also an adavntage for Crowdin's views. On Crowdin you can also see like 20 strings on screen while Weblate can only show like 10 because one string takes so much space (especially vertical space) in comparison (these numbers were made up but the difference is large). Sure, this cuts when the page changes, and occasionally it happens just when you'd like to see a few strings behind, but not that often.

Weblate is also too quick to save strings automatically if you use mouse to add variables for example - I prefer to hit Ctrl Enter instead of tab to save. Not that I don't know that I can add variables from the keyboard too but it is common to browse through and find individual strings depending on what I am doing at the time, and for extremely context-dependent language like Finnish where direct translation just simply doesn't work for a lot of the time, I usually leave obscure or uninformative strings last. The popup notifications also fill your screen in seconds if you post a few translation in a row with issues (rare, but not that rare, especially when you work with a language like Finnish where we must inflect eveything, includin names, for example). Then you don't see anything until you either manually clear all notifications or wait for them to expire (the delay is needlessly long, at least for common ones). Much of this comes down to individual preference and I am extremely pro customization and would like to see all services offer much more varied editing views.

As for free/freedom. I was kind of expecting something like that, and for the most part, I agree. BUT the problem with open source is that unless there is a party that consists of more than just a couple of enthusiasts, they tend to be cool for a while and then they die out unless there is an actual organization behind them. I have also found the communities to often be kind of blunt and limited in that they expect everyone to contribute and start adding to the clde without even bothering to consider a users level or their priorities - so it is very common to see comments along the lines of: "If you want it to be fixed, do it yourself, I don't have the time". Or people get a bunch of links to recouces they have no understanding of nor desire to learn, and when they bring this up that same attitude rises its head again. Obviously I'm generalizing, and this has nothing to with FDM which is an old project, but that is a super common attitude, and it is often also considered somehow offensive when someone then replies that they don't have time to dedicate to it. So, very often that freedom is though of as the user being able to fix, build and change things themselves. Which is cool except there are plenty more people who want to have anything to do with the code, and if the project does not want to openly support those, then it will forever remain niche for the select few and eventually dies out when they lose interest (or just grow old). In the commercial side, money kind of removes these arguments. Sure there are plenty of great os projects too, like Proton's products (Mail, VPN, Calendar, Pass...) where I've been the Finnish language lead for a while (thankfully they happen to use Crowdin Enterprise - otherwise I probably would have never gotten so deeply involved with it).

As for terms and license stuff, as a general user, I just don't care much. Those are things for the developer to worry about. I don't care what license my translation are under as I give them away freely. Do whatever you want with them, I don't mind. I do this for the common good, not for my own benefit, and if I wanted compensation or recognition, I would work as a freelancer (not that I mind either but those are just perks that some larger projects occasionally reward you with). As for licensing in general, people have recently learned that entities like Steam (as a visible example) actually sell just a license to use stuff instead of actual ownership, and what a shocker! this has been for people. People in general are kind of stupid, and/or oblivious to what they are doing - it took only 20 something years for them to suddenly realize this about Steam although this has always been the case, with almost any product, physical, digital, closed source, open source. Even super ole games/apps that are on physical disks are still the property of their publishers and even if you have purchased such a disk, you aren't allowed to do whatever you wish to its contents, just like you aren't necessarily allowed to destroy a piece of art even if you "own" it. In the end it is not that different from open source licenses which dictate the same things, just with different sets of rules. And most general users just don't care one way or the other, until they actually end up suffering from them somehow. I fall in between in that I am fully aware of the differences (or potential differences) but mostly I too just don't care. I would be pretty surprised if Steam, this huge and successful commercial entity, were to fall of the map along with my hundreds of game licenses, although technically their terms kind of allow them to stop distributing the games. Many years ago I did actually lose access to my account due to a data breach where my then duplicated password was leaked, and I would have lost the account, plain and simple, unless I could prove that I was mine. And this is something that people fail to understand about digital stuff - there is nothing that links things to you/your true identity, unless you provide such information, or keep sone token that can be used for verification, like a few physical linked to my Steam account for example. This is all stuff why licenses exist and they define who is respinsible of what, and what the parties are allowed to do, and people should really learn (or rather, bother) to do at least some research and thinking for themselves. On the other hand if a smallish open source project like FDM were to lose its main devs, I wouldn't be surprised at all (people don't live forever). And if there aren't an actual organization behind it, keepung it afloat and managing things at leaat to some extent, commercial or non commercial, it would likely die out. Commercial businesses tend to get sold and their products, if popular, tend to stay around even afterwards. That is "total freedom" vs. commercial recources for you... Businesses buy MS Office because they want that support and safety of a known entity, although LibreOffice woyld work just fine for most. Not taking into account all those Office veterans who I can already hear shouting in the background, and who are actually kind of the main reason why this will also never change - that lot just doesn't want to and won't lear anything other than what they have used for decades, and licenses and their rights are irrelevant as long as it just works like it has always done. And that right there is the main force that drives the masses. 🙂

As for Crowdin, it allows tunning a basic and small project for free and FDM is itty bitty, super tiny, atleast when it comes to string count. Also depending on the projects income, they might happily pay for a commercial service too if it happens to fit their needs. This is not my problem and as a translator I just gave my opinion from a long time user's perspective. Terms and costs are the project managers problems and whatever works for their code and their values. FDM's string count is so small that the platform is kind of irrelevant anyway, although they could then add the website there too... I just personally like to use Crowdin way more than any other platform, Pontoon included, and I just care about the workflow and usability.

**Comment by comradekingu** [comment]
You say Crowdin gives you a place to cut.
What exactly are you cutting other than your workflow?
If your workflow somehow doesn't include going back for anything but the arbitrary amount,
it isn't something I have ever seen done to any great effect.

You can adjust how many strings are loaded in Weblate, try it.

When you say you see 20 vs. 10, that is a topic to explore.
The numbers are wrong, and the comparison is further erroneous.

On Crowdin you are actually seeing just the one string, with its full translation.
The rest is an overview of strings, in only their source-string form.
Then you as a result of this taking up space, get an editing area that is shorter.

This means strings and translation line-wrap more often,
and also wrap at different places because the widths differ between languages.

You may like this way of doing things, or the side-by side view in Weblate, but your eyes don't.

It is _way_ slower, and with complex text it is an unbelievable pain.
In turn you get translations that aren't completed, which is what Crowdin has.
Over time this leads to multiple translators doing different things, with no overview of many strings, as described.

If you are lucky enough to have to go back to check on something within the limitation, you then have to
span across the screen for no good reason.
It is a broken UI for a translator workflow, invariably resulting in garbage.

Complaining about the model in a project that uses it seems unreasonable.
Moreover the critique doesn't apply to Weblate either.
It _does_ however apply to Crowdin, where unscrupelous methods are used and more to the point,
the terms and services are egregious.

Yes, it can be down to only developers to care. Those people matter.
Not having broken locales matter.
Making decisions based on most people not caring, or caring 20 years down the road,
I guess you are just one of those people.

If you say your choice doesn't matter to you, it doesn't equate giving it as advice doesn't matter to anyone else.

In summary, you can complain about the expectation and ability to do something productive about reporting bugs in Weblate, but for Crowdin are never fixing neither the UI, the model, nor the terms. It is all junk, by design.
It is a failing model because people caught on to _why_ it is that someone would hide their source code to begin with.

>Commercial businesses tend to get sold and their products, if popular, tend to stay around even afterwards. 

How does taking away anyone's right to do something about it do more to ensure nothing is taken away?
Guess how many commercial companies exist 5 years after being founded.

You get exactly what Crowdin dictates, at all times. I am elated it is rejected at large by the libre software world.

**Comment by olavinto** [comment]
I mean that I am fine with Crowdin splitting the strings every 50 strings (which is what it shows/page). I kind of feel like every page change is a milestone where I know that I've now done 50, 100 or whatever strings and I really don't mind it. If you change Crowdin's default string sorting and/or what strings are shown, you do not lose your position/get that "jump" when the page changes. Crowdin's search and filtering also works much faster. Weblate's is fine too and in a way it is a bit more powerfull with many different selevtors in the UI but I have had very little use for those. Crowdin's search is simpler and you can just directly search for anything and as default it gives you results from source and target strings, and keys.

I don't know, I think that I can see quite many strings at once one Crowdin (and this was just now taken from a phone so on desktop the view gives a lot more content): 
![Screenshot_2024-12-30-03-01-52-18_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/user-attachments/assets/a3884dc8-c537-4c7b-80de-7abcb621f311)

It is also a fact that the eyes generally like reading narrower texts than wide ones which is why magazines, websites and other text heavy publications are usually split into columns or other ways.

I do a lot of maintenance and go through old translation and strings, translated by me and others, and I have a very high quality target for myself. If line wraps are an issue then the source project has not taken this into account in its UI and this will usually be a problem anyway regardless of what you do. Crowdin shows you character count all the time as source/translation so you can easily see where you are at. In Finnish we can have long compound words and can have like five different forms for individual words, and projects not offering basic stuff like separate plural forms or decent/adaptive spaces for the texts is often a problem, especially on smaller projects that aren't really planned with localization in mind. I'd argue that setting string limits, offering contextual clues and just plain developing with these things in mind should be a priority for any devs that really value localization. Crowdin sure has its faults too but generally I find it to be much much faster for me to use than anything else, including maintaining older strings and consistency.

Smaller open source projects or other "homegrown" projects simply tend to lack the recources to really get into this to do any of those. I don't know about you but I have found that many Crowdin users aren't familiar with their UI and don't know that they have various ways to change the layout, sorting and filtering. I'm also a proofreader on most projects that I actively translate so I may have a bit more access to some features. Some are also dependant on what the project actually has enabled.

Me not reporting Weblate bugs is true, but it doesn't change the fact that they are there. I just stated that they are there, and something like the zen mode's side by side view's column width not staying fixed should be pretty easy to spot if one does any testing at all. I also don't really like to post simple bug reports without doing some wider testing and data gathering myself first, with screenshots and possibly even videos, nor do I generally have time to followup and test fixes. I mostly do get to reportings only when they start to really get on my nerves. And as I work as a maintenance technician, I really don't feel like fixing and troubleshooting other problems after work.

Finally, I am not really giving advice here, just my opinion based on my experience. And I've already stated that Weblate is fine.

For me software being open or closed source is almost completely irrelevant. I don't use things just because I can see under the hood and I'm fine with just being able to use it. What values and principles the devs have is a different thing and closed source or moneytizing one's work is not necessarily a bad thing. Also the management side doesn't really concer me as I am not managing, in this case FDM, and they shoyld obviously test drive these services and see what fits best for them. I would happily use any of these instead of editing the files directly as it works now. Not that Poedit is bad, I just don't care to manually transfer files around these days.

I am just a translator/proofreader and I have translated and maintained many projects for many years, even before there was any tools like these. I really have very little interest or energy to dig through code or write any myselft. I used to build websites but not really for many years now (not counting maintaining one small business site for a friend's company).

If Crowdin doesn't fit you, don't use it. I use Weblate, Transifex and a few others because some devs have chosen those. And whatever FDM chooses, I will use.

A few more just:
![Screenshot_2024-12-30-04-59-53-66_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/user-attachments/assets/21babd78-7d36-4325-b563-2a734f9567ea)
![Screenshot_2024-12-30-04-59-08-56_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/user-attachments/assets/04c81141-4a7f-4a0b-9ac3-44113475181a)
![Screenshot_2024-12-30-05-05-14-87_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/user-attachments/assets/5a38a3b1-bec6-47a4-8e7d-83aed9c0bb9d)

**Comment by comradekingu** [comment]
That something doesn't bother you doesn't take away from
the nature of the problem of wasting time for every single user.

Catching problems in other languages is just making the most of eyeballs present.
Missing out on it is a big deal.

The human mind natively compartmentalises information in hierarchies of distances and space.
Making that sort of navigation impossible and at best cumbersome is a default loss of capacity.

Yes it is quicker to read narrower text, but you are misapplying the findings.
The actual problem is going between texts where the lineshifts are now as arbitrary as the length of translations between languages.
To add insult to injury the two are in the wrong plane, and/or redundant.

What you want is to represent the lineshifts actually present in the text, which is possible at the greatest possible horisontal display area.
If you narrow anything down beyond that, you are failing. Narrower is worse, and it is only a factor when it fails, and it does.

Why you went out of your way to configure anything to be worse is anyones guess.
It represents that same broken UI problem present in Crowdin.

Age of platform doesn't change it, character count doesn't change it, you saying "it is faster" doesn't change it.

As for configuration, by default strings get held up in the voting +1 system at Crowdin, which is how when someone knows what to do about fixing something, one ends up waiting for a third person to apply the fix.
It is a first-mover drive-by platform, where even correcting anything at a consistent level is near impossible.

As a heads up, you translated "GIT" into "GIT" when it should be "Git".

What values are present in open source being represented by values shared by devs is not even wrong, but they are present in the license, where it does matter.

**Comment by olavinto** [comment]
I still don't have any clue why you are trying to force me to do things your way. Most strings fit side by side in one line (unless you have a super narrow screen) and longer ones tend to be descriptions of some kind that just wrap freely anyway (unless the apps UI is locked and doesn't expand/contract with text, which is what most apps do these days). But telling me that my approach is just wrong, is just offensive. I have never called your approach wrong and I have absolutely nothing against you (or others) preferring wide (or whatever) views, or disliking narrower one's for that matter. That would just be super narrowminded, trying to fit everyone into the same box.

Usually most strings have no line changes and the UI's wraps them as needede, and if you really need to aim close to the lenght of the source, character count does get you there and if that isn't enough, then the source is not very localization friendly. And if the UI does't allow strings of certain lenght, max character counts can be defined and there are many projects that do this. Sure, occasionally there are situations where lenght is crucial but those are rare and I can check them out by simply widening the views. These are generally like needles in a haystack. If these are constantly present, I will probably not get involved such a project because then there are always tons more problems where we can't use correct forms and are forced to ymuse unnatural expressions to avoid things and I do not have time nor interest to try to make those kinds of projects fit the requirements of my languge. It is also a very narowsighted approach if you think that every translation should ve of equal lenght with the source. That makes translations often impossible and depending on your target language even single words might have twice or even longer lenght of the source (in Finnish this may happen sometimes). All of this is something that developers should take into account with their UI design to make it localization friendly so that we don't have to get stuck trying to somehow force our languages to fit into a static, limited frame.

I have spent a huge amount of time proofreading the projects I do it for, including checking strings against the actual UI's and making sure that they fit physically and that they are contexually correct. So the argument that I have failed or keep failing is completely outrageous and disrespectful. Also to be more accurate, this is about localization, not about translating, which means more than just trying to match the content of the source string. This also means that there may be differences between the source and target strings that a previous localizer has deemed reasonable for the context and if someone can just come and override those with direct translations without any oversight, that isn't good either. Finnish is a good example where this is sometimes very beneficial and even necessary, and very often we can't just translate strings as they are because our language is very dependant on source context. And this is super hard for some projects to understand. Just recently I asked context for one project and they told me to just translate them as they are because it wil be better than nothing anywya - a bad attitude because that easily produces completely inaccurate translations that may speak of something completely different (in this case they could just as well use AI translation and call it good).

Botw. GIT has now been changed into Git in the translation, thanks for reporting. Technically you can use GIT too, if you wish to emphasize it for some reason which is probably why the source has completely capitalized it. I am sure that there are many more mistakes to find, I am not perfect. I am, however, extremely satisfied of the level Kodi, and many other projects are after I've gone through them, and I know how much work and dedication I've put into them (and this includes checking in their actual emvironment by actually using the end products). Even if I have haved mostly used a side by side view to do it... 🤷
